backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        /// <returns></returns>
+        public ProjectReport Add(ProjectReport report)
+        {
+            if (report.To == null)
+            {
+                report.To = DateTime.UtcNow;
+            }
+            var mostRecentReport = this.Context.ProjectReports
+                .Where(r => r.ReportTypeId == ReportTypes.SPL)
+                .OrderByDescending(r => r.To)
+                .FirstOrDefault();
+            report.From = mostRecentReport?.To;
+            this.Context.Add(report);
+
+            this.Context.ProjectSnapshots.AddRange(GenerateSnapshots((DateTime)report.To));
+
+            this.Context.CommitTransaction();
+
+            return report;
+        }
+
+        /// <summary>
+        /// Update the specified project report metadata in the datasource.
+        /// If the project 'To' date has changed, 
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="RowVersionMissingException">Project report rowversion is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project does not exist.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to edit report.</exception>
+        /// <returns></returns>
+        public ProjectReport Update(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var originalReport = this.Context.ProjectReports
+                .SingleOrDefault(p => p.Id == report.Id) ?? throw new KeyNotFoundException();
+
+            if (originalReport.To == null) throw new ArgumentNullException();
+            if (originalReport.IsFinal && report.IsFinal) throw new InvalidOperationException($"Unable to update FINAL project reports.");
+            if (originalReport.From == originalReport.To) throw new InvalidOperationException($"Project report start and end dates cannot be the same.");
+
+            this.Context.Entry(originalReport).CurrentValues.SetValues(report);
+
+            //If the report 'To' date has changed regenerate all the snapshots based on that date.
+            if(!originalReport.To.Equals(report.To))
+            {
+                this.Context.ProjectSnapshots.AddRange(GenerateSnapshots((DateTime)report.To));
+            }
+
+            this.Context.SaveChanges();
+            this.Context.CommitTransaction();
+
+            return Get(originalReport.Id);
+        }
+
+        /// <summary>
+        /// Generate an updated list of snapshots based on current project data.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="NotAuthorizedException">User does not have permission to get refreshed snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> Refresh(int reportId)
```

Comment:
I think this ideally would be returning a new `ProjectReport` as the `To` date will change.  However because the relationship is loosely defined with a date that wouldn't provide the appropriate information.  Not sure anything needs to change here.